### ISSUE TITLE
NonNullable FormControl is assignable to Nullable FormControl, leading to deceptive type safety

### ISSUE THREAD

**Issue by pburgmer** [angular/angular#48727]
### Which @angular/* package(s) are the source of the bug?

forms

### Is this a regression?

Yes

### Description

Use case: a component with an input of type `FormControl<boolean | null>`. User is able to assign a `FormControl<boolean>`. Component can internally set null as value. User receives an unexpected value at runtime but no compiler error.

```
const nonNullable = new FormControl(false, {nonNullable: true});
const nullable: FormControl<boolean | null> = nonNullable;
nullable.setValue(null); // nonNullable control receives value null which it does not expect;
```

Per default typescript does not handle variance / restrict variance. TypeScript 4.7 introduced [variance annotations](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#optional-variance-annotations-for-type-parameters). Marking TValue in FormControl as invariant solves the problem in my reproduction test. But I don't know all the effects this change would have applied in Angular.

### Please provide a link to a minimal reproduction of the bug

https://www.typescriptlang.org/play?ts=4.7.4&ssl=17&ssc=122&pln=17&pc=4#code/PTAEGcEsFsFcBsCGAXApuUBBAdgcwYgE4YDGshhq2yoMADvKtFcipAPbagAUAbgIwAWAJQAoEknAYAYu0LQAwp2SF28ADwAVAHygA3qNCgSncCtglkc7nUKReKVKAfxYqAFyhNw-QF9DEKjIAGqIrqh8YW6e3vqgyAAWkOAAdC5uoAC8zlGo-v7ipjTYnAByCEgARoxZoNioAO6gsvJK1Krw3ABmYeCowgDcoiXY5fBVjCl9Ibnc5v1DIKAAAsjgALSoAB50qJabFHJ1nOvYFYjVTl1y0MbKHRAJ7AgAJsc0iCQkqHQ06U5ncbDMrnS5TIKhcLcQHwQaiURLVYbba7faoQ6EY7YU6gmrXeR3dpqR7PeBvEo0SpORBSSC4bAXGrsRLohrJJyQGgmV6gXBBOoVQIfDD-QrYMwC8aMjzNG5tFRqdSVdhqVCILgAH0l8F02RGYwmqCGMOl4JmUJhg1AS31uKcJiJ8FAlG+9nQOXC2tADSSJAStBoL3Y7opoBRe2QA3iCSclDCoFs7Eutx9kD9JJ5VOMKD9qDe4HYzB96NQ8KW4CemaukC20u9nP9Djs6u+oHVFLYpgRYASyGQdHA7hAL1QvGq7FwqWgadUBa6yBSJmgwGQAE9duASHZfsB289sCRIHh1muN1vIL91oJ1gB2YAAYnYvw4DPg6ybkBbqHWe9YyBfGz4ie67fnQRCIMwaDEN2oCMMgADkpAxiQADWZZgJooAJDS8TsKAWbQEQKF5m2GBHh+6o0FYCY0hgYROmgZjgOIkgYAAktgFHUC0ij3IqR6gM8NA6PoAQOmYhAWFYhA2HYDhoB60ReD4ej+EY0yQm4kThDEKnRskaS5LU-z5PC4nFCCUqXBxXE0HqjSgDZRCftxcp8Z0PTwH0cK2lZjBOc21BmppETzHCiJrJsOwRgcqiYiMOJ+VcNyEgqTpBiGzJtl8Px-EZMLAqMdoBS5C4abMlpDDBSJRaiyCxUcCUmpcoD4rcDppRmZLvPh1K0vSdbMjGhBsn0AZ3DyfLFIK0ykYppbmdq0olZRngra5rTuUqKqMOqoBajCupYgay2cc5lHGsVZ2BWVEIVRUcIweqbxbuwUjGMhKH1ok+E7WqmramKErmPAq4ndZ12lbU9RNOtyA8fKHTbaqe0HRU2jdL0CyiCDYNXbZwWzGFQy4+D-mQ5RhMWg9VVLM9bZeXhiAfXsX1sj94CwJUp5HrgLG0VgRhC8LRh6G2tTwZg8GgP4EgCwAQmGWxoNgLwYJgcSVBL8vS7LrGgAoSsq2roCK2LJASwoutmUU4v2U0mDcHCi1a-bptO0Mi0W27Cgezb4o0Jz3Prrz0MOXDCNbfLGOVOFYA1eGaIYoJ2Cg-hbaqxAXOnu69HsA0eYADTHFnuyYjnohBzzeBU1piBwlXIc1+VUIkIMQA

### Please provide the exception or error you saw

```true
The error is: no compile time error for assigning the nonNullable control.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular 14.2
```

**Comment by JeanMeche** [comment]
`nonNullable` has no runtime implication outside of the `constructor` where the `defaultValue` is set to the initial value. 

As you could see with this valid unit test:
```
    it('supports invariance', () => {
      const nonNullableCtrl = new FormControl(false, {nonNullable: true});
      expect(nonNullableCtrl.valid).toBe(true); // OK

      //@ts-expect-error nonNullable can't be assign to nullable.
      const nullableCtrl: FormControl<boolean | null> = nonNullableCtrl;
      nullableCtrl.setValue(null); 


      expect(nullableCtrl.valid).toBe(true); // ok
      expect(nullableCtrl.value).toBe(null); // ok
    })
  });
```

I'm not sure this feature would be welcome.

**Comment by pburgmer** [comment]
Hi Jean, 

Thanks a lot for the quick response. I'm confused now. These two lines of your code snippet do exactly what I'm asking for.
```
//@ts-expect-error nonNullable can't be assign to nullable.
const nullableCtrl: FormControl<boolean | null> = nonNullableCtrl;
```
the ts-expect-error comment (and the assumption that this test passes) tells us that it behaves in your setup like I wish it would in my setup. The assignment of a nonNullable control to a nullable should not be possible. But in my project and in the reproduction stackblitz it is assignable. Please see line 17 in the reproduction example. That is exactly my problem. It is assignable but imho it shouldn't.

I don't expect any runtime checks. I just wish that the compiler does not tell me everything is ok ;)

**Comment by JeanMeche** [comment]
Yeah sorry for that disabling comment, I did try implementing your solution which works well (and doesn't break any unit test in the repo by the way). Beside of that, my comment stands. 

To me, this is up to the typings, only the typings decide what behaviour you're expecting so this is more a feature than a problem to me. 

I will let the maintainers and come back with a "official" anwser/decision tough.

**Comment by pburgmer** [comment]
I'm still wondering why the test compiles. If there is a ts-expect-error comment but the compiler does not detect an error on the next line that combination should lead to an error itself, shouldn't it?

**Comment by JeanMeche** [comment]
Like i said, I did implement the invariant solution before implementing this test. That's why it worked ! 

Btw, I didn't know this feature, thanks for the heads-up !
